ocaml-lsp-server/src/ocaml_lsp_server.ml

```diff
@@ -656,33 +656,205 @@ let text_document_lens (state : State.t)
 
 let folding_range (state : State.t)
     { FoldingRangeParams.textDocument = { uri }; _ } =
-  let+ outline =
-    let command = Query_protocol.Outline in
-    let doc = Document_store.get state.store uri in
-    Document.dispatch_exn doc command
+  let folding_range (range : Range.t) =
+    FoldingRange.create ~startLine:range.start.line ~endLine:range.end_.line
+      ~startCharacter:range.start.character ~endCharacter:range.end_.character
+      ~kind:Region ()
   in
-  let folds : FoldingRange.t list =
-    let folding_range (range : Range.t) =
-      FoldingRange.create ~startLine:range.start.line ~endLine:range.end_.line
-        ~startCharacter:range.start.character ~endCharacter:range.end_.character
-        ~kind:Region ()
-    in
-    let rec loop acc (items : Query_protocol.item list) =
-      match items with
-      | [] -> acc
-      | item :: items ->
-        let range = Range.of_loc item.location in
-        if range.end_.line - range.start.line < 2 then
-          loop acc items
-        else
-          let items = item.children @ items in
-          let range = folding_range range in
-          loop (range :: acc) items
+  let ranges = ref [] in
+  let push (range : Range.t) =
+    if range.end_.line - range.start.line >= 2 then
+      ranges := folding_range range :: !ranges
+    else
+      ()
+  in
+  let fold_over_parsetree (parsetree : Mreader.parsetree) =
+    let iterator =
+      let default_iterator = Ast_iterator.default_iterator in
+      let structure (iterator : Ast_iterator.iterator) structure =
+        List.iter structure ~f:(fun structure_item ->
+            iterator.structure_item iterator structure_item)
+      in
+      let signature (iterator : Ast_iterator.iterator) signature =
+        List.iter signature ~f:(fun signature_item ->
+            iterator.signature_item iterator signature_item)
+      in
+      let type_declaration (_iterator : Ast_iterator.iterator)
+          (type_declaration : Parsetree.type_declaration) =
+        let range = Range.of_loc type_declaration.ptype_loc in
+        push range
+      in
+      let module_type_declaration (iterator : Ast_iterator.iterator)
+          (module_type_declaration : Parsetree.module_type_declaration) =
+        let range = Range.of_loc module_type_declaration.pmtd_loc in
+        push range;
+        match module_type_declaration.pmtd_type with
+        | None -> ()
+        | Some module_type -> iterator.module_type iterator module_type
+      in
+      let module_type (iterator : Ast_iterator.iterator)
+          (module_type : Parsetree.module_type) =
+        match module_type.pmty_desc with
+        | Parsetree.Pmty_ident _ -> ()
+        | Parsetree.Pmty_signature signature ->
+          iterator.signature iterator signature
+        | Parsetree.Pmty_functor (_, _) -> ()
+        | Parsetree.Pmty_with (_, _) -> ()
+        | Parsetree.Pmty_typeof _ -> ()
+        | Parsetree.Pmty_extension _ -> ()
+        | Parsetree.Pmty_alias _ -> ()
```

Comment:
1. cases which do the same thing can be collapsed, eg `| Pmty_typeof _ | Pmty_extension _ -> ()` 
2. we can keep the `Parsetree.` prefix only for the first case and remove for others - it hints the compiler where to look for the value constructor